pkg/agent/clientset.go

```diff
@@ -246,3 +273,72 @@ func (cs *ClientSet) shutdown() {
 		delete(cs.clients, serverID)
 	}
 }
+
+func (cs *ClientSet) syncPodStatus() {
+	if !cs.enablePodCondition || cs.k8sClient == nil || cs.podNamespace == "" || cs.podName == "" {
+		return
+	}
+
+	pod, err := cs.k8sClient.CoreV1().Pods(cs.podNamespace).Get(context.TODO(), cs.podName, metav1.GetOptions{})
+	if err != nil {
+		klog.Warningf("error while getting pod: %s", err)
+		return
+	}
+
+	updatedPod := cs.appendPodCondition(pod, getPodCondition(cs.ClientsCount(), cs.serverCount))
+	if updatedPod == nil {
+		return
+	}
+
+	_, err = cs.k8sClient.CoreV1().Pods(updatedPod.Namespace).UpdateStatus(context.TODO(), updatedPod, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Warningf("error while updating pod status: %s", err)
+		return
+	}
+}
+
+func (cs *ClientSet) appendPodCondition(orig *corev1.Pod, condition *corev1.PodCondition) *corev1.Pod {
+	for i, cur := range orig.Status.Conditions {
+		if cur.Type != condition.Type {
+			continue
+		}
+
+		if cur.Status == condition.Status &&
+			cur.Reason == condition.Reason &&
+			cur.Message == condition.Message {
+			return nil
+		}
+
+		pod := orig.DeepCopy()
+		pod.Status.Conditions[i] = *condition
+		return pod
+	}
+
+	pod := orig.DeepCopy()
+	pod.Status.Conditions = append(pod.Status.Conditions, *condition)
+	return pod
+}
+
+func getPodCondition(current, desired int) *corev1.PodCondition {
+	condition := &corev1.PodCondition{
+		Type:               connectedPodConditionType,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	}
+
+	if current == 0 || desired == 0 {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = disconnectedPodConditionReason
+		return condition
+	}
+
+	if current < desired {
+		condition.Status = corev1.ConditionFalse
+		condition.Reason = partiallyConnectedPodConditionReason
+		condition.Message = fmt.Sprintf("connected to %d of %d servers", current, desired)
```

Comment:
should we find a way to represent specific client errors? It would be useful to have at least a high-level indication of the client error e.g. connection timeout, dns can't resolve, etc. Although this may be difficult to implement.